File: agent/library/src/main/java/org/matilda/commands/listener/CommandResponseListeningInstance.java
package org.matilda.commands.listener;

import org.matilda.commands.CommandFailedException;
import org.matilda.commands.protobuf.CommandResponse;
import org.matilda.messages.Message;
import org.matilda.messages.listener.MessageListeningInstance;

import java.io.IOException;


public class CommandResponseListeningInstance implements AutoCloseable {
    private final MessageListeningInstance mMessageListeningInstance;

    public CommandResponseListeningInstance(MessageListeningInstance messageListeningInstance) {
        mMessageListeningInstance = messageListeningInstance;
    }

    public byte[] waitForResponse() throws InterruptedException, IOException, CommandFailedException {
        Message message = mMessageListeningInstance.waitForMessage();
        CommandResponse commandResponse = CommandResponse.parseFrom(message.data);
        if (!commandResponse.getSuccess()) {
            throw new CommandFailedException(new String(commandResponse.getResult().toByteArray()));
        }

        return commandResponse.getResult().toByteArray();
    }

    public void stop() {
        mMessageListeningInstance.stop();
    }

    public void close() {
        stop();
    }
}

File: agent/library/src/main/java/org/matilda/commands/listener/CommandResponseListeningInstance.kt
package org.matilda.commands.listener

import org.matilda.commands.CommandFailedException
import org.matilda.commands.protobuf.CommandResponse
import org.matilda.messages.listener.MessageListeningInstance
import pbandk.decodeFromByteArray
import java.io.Closeable

class CommandResponseListeningInstance(private val mMessageListeningInstance: MessageListeningInstance) : Closeable {
    suspend fun waitForResponse(): ByteArray {
        val message = mMessageListeningInstance.waitForMessage()
        val commandResponse = CommandResponse.decodeFromByteArray(message.data)
        if (!commandResponse.success) {
            throw CommandFailedException(String(commandResponse.result.array))
        }
        return commandResponse.result.array
    }

    private fun stop() {
        mMessageListeningInstance.stop()
    }

    override fun close() {
        stop()
    }
}
